x/genutil/types/genesis.go

```diff
@@ -0,0 +1,157 @@
+func (ag *AppGenesis) ToCometBFTGenesisDoc() (*cmttypes.GenesisDoc, error) {
+	var consensusParams *cmttypes.ConsensusParams
+	if ag.ConsensusParams != nil {
+		consensusParams = &cmttypes.ConsensusParams{
+			Block: cmttypes.BlockParams{
+				MaxBytes: ag.ConsensusParams.Block.MaxBytes,
+				MaxGas:   ag.ConsensusParams.Block.MaxGas,
+			},
+			Evidence: cmttypes.EvidenceParams{
+				MaxAgeNumBlocks: ag.ConsensusParams.Evidence.MaxAgeNumBlocks,
+				MaxAgeDuration:  ag.ConsensusParams.Evidence.MaxAgeDuration,
+				MaxBytes:        ag.ConsensusParams.Evidence.MaxBytes,
+			},
+			Validator: cmttypes.ValidatorParams{
+				PubKeyTypes: ag.ConsensusParams.Validator.PubKeyTypes,
+			},
+		}
+	}
+
+	return &cmttypes.GenesisDoc{
+		ChainID:         ag.ChainID,
+		InitialHeight:   ag.InitialHeight,
+		AppHash:         ag.AppHash,
+		AppState:        ag.AppState,
+		ConsensusParams: consensusParams,
+		Validators:      ag.Validators,
+	}, nil
+}
+
+// SaveAs is a utility method for saving AppGenesis as a JSON file.
+func (ag *AppGenesis) SaveAs(file string) error {
+	appGenesisBytes, err := json.MarshalIndent(ag, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(file, appGenesisBytes, 0600)
+}
+
+// ValidateAndComplete performs validation and completes the AppGenesis.
+func (ag *AppGenesis) ValidateAndComplete() error {
+	if ag.ChainID == "" {
+		return errors.New("genesis doc must include non-empty chain_id")
+	}
+
+	if len(ag.ChainID) > MaxChainIDLen {
+		return fmt.Errorf("chain_id in genesis doc is too long (max: %d)", MaxChainIDLen)
+	}
+
+	if ag.InitialHeight < 0 {
+		return fmt.Errorf("initial_height cannot be negative (got %v)", ag.InitialHeight)
+	}
+
+	if ag.InitialHeight == 0 {
+		ag.InitialHeight = 1
+	}
+
+	if ag.GenesisTime.IsZero() {
+		ag.GenesisTime = cmttime.Now()
+	}
+
+	// verify that consesus and validators parameters are valid for CometBFT
+	// TODO eventually generalize this for every consensus engine
+	cmtGenesis, err := ag.ToCometBFTGenesisDoc()
+	if err != nil {
+		return err
+	}
+
+	if err := cmtGenesis.ValidateAndComplete(); err != nil {
+		return err
+	}
+
+	ag.Validators = cmtGenesis.Validators
+	consensusParams := cmtGenesis.ConsensusParams.ToProto()
+	ag.ConsensusParams = &consensusParams
+
+	return nil
+}
+
+// AppGenesisFromFile reads the AppGenesis from the provided file.
+func AppGenesisFromFile(genFile string) (*AppGenesis, error) {
+	jsonBlob, err := os.ReadFile(genFile) //nolint:gosec
+	if err != nil {
+		return nil, fmt.Errorf("couldn't read AppGenesis file (%s): %w", genFile, err)
+	}
+
+	var appGenesis AppGenesis
+	if err := json.Unmarshal(jsonBlob, &appGenesis); err != nil {
+		// fallback to CometBFT genesis
+		var ctmGenesis cmttypes.GenesisDoc
+		if err2 := cmtjson.Unmarshal(jsonBlob, &ctmGenesis); err2 != nil {
+			return nil, fmt.Errorf("error unmarshalling AppGenesis at %s: %w and failed fallback to CometBFT GenDoc: %w", genFile, err, err2)
```

Comment:
just asking, can we use more than one error wrapping directive in `fmt.Errof`, doesn't it error while testing?
I remember something similar with my work a way back, not really sure.
Idk, but maybe you could use string for one of the error wrappers.
```suggestion
			return nil, fmt.Errorf("error unmarshalling AppGenesis at %s: %s and failed fallback to CometBFT GenDoc: %w", genFile, err.Error(), err2)
```

Reply:
Yes, we can since Go 1.20 :)